cpu/cc2538/syscalls.c

```diff
@@ -0,0 +1,60 @@
+/*
+ * Copyright (c) 2014, Institute for Pervasive Computing, ETH Zurich.
+ * All rights reserved.
+ *
+ * Author: Andreas Dröscher <contiki@anticat.ch>
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ * 3. Neither the name of the Institute nor the names of its contributors
+ *    may be used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE INSTITUTE AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED.  IN NO EVENT SHALL THE INSTITUTE OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+ * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
+#include <sys/stat.h>
+#include <string.h>
+
+#define DEBUG 0
+#if DEBUG
+#include <stdio.h>
+#define PRINTF(...) printf(__VA_ARGS__)
+#else
+#define PRINTF(...)
+#endif
+
+static char *allocated = 0;
+
+caddr_t _sbrk(int incr) {
+  extern char _heapdata;  // symbols are defined in cc2538.lds
+  extern char _eheapdata; // they mark end of bss and end of memory.
+
+  if (allocated == 0) {
+    memset(&_heapdata, 0, (unsigned int)&_eheapdata - (unsigned int)&_heapdata);
+    allocated = &_heapdata;
+  }
+
+  if (allocated + incr > &_eheapdata) {
+    PRINTF("Out of heap space!\n");
+    return (caddr_t)0;
```

Comment:
I have checked: I confirm that Newlib needs `-1` here.

As to `errno`, it needs something, but it's more complicated, because the `errno` from `errno.h`is not the same as the `errno` used by `_sbrk_r()`. See https://sourceware.org/git/gitweb.cgi?p=newlib-cygwin.git;a=blob;f=newlib/libc/include/sys/errno.h;h=a72c37320a7681ae28da520ad719139236e378fc;hb=HEAD vs. https://sourceware.org/git/gitweb.cgi?p=newlib-cygwin.git;a=blob;f=newlib/libc/reent/sbrkr.c;h=44e42b57f6ad0d6f395fe29bf6a1294fc64f4462;hb=HEAD. The former uses the function from https://sourceware.org/git/gitweb.cgi?p=newlib-cygwin.git;a=blob;f=newlib/libc/errno/errno.c;h=fd1743d73625bf098fc23af9c06924381f840139;hb=HEAD to access the `errno` field inside `_REENT`, while the latter uses a global `errno` variable. To make this work, we have to do like `_sbrk_r()` since we are constrained by it, even if this is not very portable, but we are already implementing a Newlib-specific function, so it's not that bad. This would also probably require to add a comment. I update my suggested code at the bottom accordingly.